### PR TITLE
Add Apache Http dependencies as needed where usage was transitive previously

### PR DESCRIPTION
**Pull request by Laurens-W**
## What's changed?

Changed from an Change Dependency structure to Add/Remove Dependency

## What's your motivation?

Transitive dependencies are not taken into account by `ChangeDependency` recipe.

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

**Comment by Laurens-W** [comment]
I can add some test cases to prove the situation at hand in a bit

File: src/test/java/org/openrewrite/apache/httpclient5/UpgradeApacheHttpClient5Test.java [hunk 1]
```diff
@@ -172,8 +173,11 @@ void migrateDependenciesWhenTwoBecomeOne() {
               </project>
               """,
             spec -> spec.after(pom -> {
-                Matcher version = Pattern.compile("5\\.4\\.\\d+").matcher(pom);
+                Matcher version = Pattern.compile("5\\.\\d+\\.\\d+").matcher(pom);
                 assertThat(version.find()).describedAs("Expected 5.4.x in %s", pom).isTrue();
```
Comment (jevanlingen) [inline src/test/java/org/openrewrite/apache/httpclient5/UpgradeApacheHttpClient5Test.java:177]: Now that you changed the pattern to `"5\\.\\d+\\.\\d+"`, you should probably change the message to `"Expected 5.x.x in %s"` as well.

**Comment by ryan-hudson** [comment]
> Added a test to show that we now unconditionally add dependencies, which is not great seeing how this upgrade is tied into larger Spring Boot migrations, meaning even projects that do not use the Apache Http client at all would see dependencies added. Typically this is then handled by `onlyIfUsing`, or some other precondition to avoid unnecessary changes.

Hey @Laurens-W @timtebeek, is there a reason we didn't implement this `onlyIfUsing` pattern as part of this PR? I'm seeing this change add httpClient5 to applications that only use AWS SDK, which is still on httpClient 4. Having the `onlyifUsing` check seems like it would help prevent additions in edge cases like this?

**Comment by timtebeek** [comment]
Hmm; from the above discussion and tests I had expected this to work using the precondition
```yaml
preconditions:
  - org.openrewrite.java.dependencies.DependencyInsight:
      groupIdPattern: org.apache.httpcomponents
      artifactIdPattern: "*"
```
@Laurens-W could you have another look at preventing unnecessary dependencies from being added?

My guess is the dependency insight triggers on a transitive, but we only need to add a dependency if used in the sources here.
